c10/core/TensorImpl.h

```diff
@@ -138,6 +138,62 @@ struct C10_API AutogradMetaInterface {
   virtual ~AutogradMetaInterface();
 };
 
+// NOTE [ Version Counter Sharing ]
+//
+// Every Tensor has a version counter. Version counters are incremented
+// whenever the data or shape of a tensor changes through Variable operations.
+// These are typically in-place operations. Version counters are used to
+// detect modifications to saved variables which would result in incorrect
+// gradient calculations. Version counters may be shared between Variables:
+//
+// 1. A view shares the version counter of the base Variable,
+// 2. Detached variables share the version counter of the source,
+// 3. Unpacked saved variables share the version counter of the source.
+//
+// Version counters are not shared when:
+// 1. We replace a `Variable`'s underlying `Tensor` by calling `set_data(...)`.
+//
+// NOTE: We explicitly don't increment the version counter in non-Variable
+// operations, because it is a useful escape hatch when we want to change the
+// content of a Variable without invalidating it in the autograd graph.
+//
+// NOTE: After the Variable/Tensor merge, a tensor will not have AutogradMeta when
+// its `requires_grad_` is false, but when we use this tensor in the forward pass of
+// a function that requires saving this tensor for backward, we need to keep track of
+// this tensor's version to make sure it's always valid in the autograd graph.
+//
+// One logical way to achieve this goal is to initialize AutogradMeta and create the
+// version counter for the non-requires-grad tensor only when it's saved for backward.
+// However, since saving a tensor for backward happens in the forward pass, and our
+// invariant is that forward pass needs to be thread-safe, lazy-initializing AutogradMeta
```

Comment:
This has nothing to do with "forward pass" specifically. We support multithreaded read access to Tensor, period. Saving a tensor is a read operation, and therefore the reasoning below follows.

Reply:
I feel that we could ask "why saving a tensor has to be a read operation", but it can't be a write operation because then the forward pass will not be thread-safe, and having forward pass work in multi-thread scenario is an important use case.

Reply:
To me, it's obvious that saving a variable for later is a read-only operation. (Forget about PyTorch variables. If I stash an object somewhere so I can look at it later, surely that "stashing" process doesn't write to the object!)

I think it's perfectly fine to give an example where making "tensor saving" a write operation would break things. But you're more apt to confuse readers if you make this front and center.